### PR TITLE
Add a function to turn Box<T> into Box<[T]>

### PR DESCRIPTION
**Pull request by elichai**
Hi,
I think this is very useful, as currently it's not possible in safe rust to do this without re-allocating.
an alternative implementation of the same function can be:
```rust
pub fn into_boxed_slice<T>(boxed: Box<T>) -> Box<[T]> {
    unsafe {
        let slice = slice::from_raw_parts_mut(Box::into_raw(boxed), 1);
        Box::from_raw(slice)
    }
}
```

The only thing that makes me a little uncomfortable is this line :
> The alignment of array types is greater or equal to the alignment of its element type

from https://rust-lang.github.io/unsafe-code-guidelines/layout/arrays-and-slices.html

But then I see:
> The alignment of &T, &mut T, *const T and *mut T are the same, and are at least the word size.
> The alignment of &[T] is the word size.

from https://rust-lang.github.io/unsafe-code-guidelines/layout/pointers.html#representation

So I do believe this is valid(FWIW it also passes in miri https://play.rust-lang.org/?gist=c002b99364ee6b29862aeb3565a91c19)

### CODE REVIEW

**Comment by Dylan-DPC** [comment]
I think it is better to have an issue about this, and also maybe add a test

**Comment by elichai** [comment]
> I think it is better to have an issue about this, and also maybe add a test

Added tests, is this what you were looking for? (I'm not sure why every error appears twice)

**Comment by cuviper** [comment]
> But then I see:
> 
> > The alignment of &T, &mut T, *const T and *mut T are the same, and are at least the word size.
> > The alignment of &[T] is the word size.
> 
> from https://rust-lang.github.io/unsafe-code-guidelines/layout/pointers.html#representation

Those are all about the alignment of pointer-like values themselves, regardless of `T`. I don't think that's relevant to your question, whether `T` and `[T]` have compatible alignment. However, [`slice::from_ref`](https://doc.rust-lang.org/std/slice/fn.from_ref.html) does testify to that.

**Comment by elichai** [comment]
> Those are all about the alignment of pointer-like values themselves, regardless of `T`. 

That makes sense. the usage of `T` there confused me (maybe we should add "or any pointer type" or something that signify it's about pointers not types) Thanks!

File: src/liballoc/boxed.rs [hunk 1]
```diff
@@ -816,6 +827,18 @@ impl From<Box<str>> for Box<[u8]> {
     }
 }
 
+#[unstable(feature = "box_into_boxed_slice", issue = "none")]
+impl<T> From<Box<T>> for Box<[T]> {
```
Comment (cuviper) [inline src/liballoc/boxed.rs:831]: AFAIK, trait impls are effectively insta-stable -- they can't really be unstable unless the trait and/or types involved are unstable.

Comment (elichai) [reply src/liballoc/boxed.rs:831]: ha, so I'll remove this from, and it can be added back if/when this is stable

**Comment by sfackler** [comment]
What's the use case for this?

**Comment by elichai** [comment]
> What's the use case for this?

I have a function that returns `Vec<syn::Type>`, by extracting a tuple into multiple types.
But if the Type is `TypeParen` it contains a `Box<Type>` which I want to return it alone in the `Vec<syn::Type>` but right now to do that I need to do `return vec![*paren.elem];` which deallocates the box and reallocates a Vec

File: src/liballoc/boxed.rs [hunk 2]
```diff
@@ -239,6 +239,17 @@ impl<T> Box<T> {
     pub fn pin(x: T) -> Pin<Box<T>> {
         (box x).into()
     }
+
+    /// Converts a `Box<T>` into a `Box<[T]>`
+    ///
+    /// This conversion does not allocate on the heap and happens in place.
+    ///
+    /// This is also available via [`From`].
```
Comment (sfackler) [inline src/liballoc/boxed.rs:247]: Is this true?

**Comment by sfackler** [comment]
Seems plausible to me. Can you make a tracking issue and update the stability annotation?

File: src/liballoc/boxed.rs (hunk 2 above)
Comment (elichai) [reply src/liballoc/boxed.rs:247]: Ops, thanks. I just removed the From impl because it's insta-stable

**Comment by elichai** [comment]
> Seems plausible to me. Can you make a tracking issue and update the stability annotation?

👍 Thanks!

File: src/liballoc/boxed.rs [hunk 3]
```diff
@@ -239,6 +239,16 @@ impl<T> Box<T> {
...
         (box x).into()
     }
+
+    /// Converts a `Box<T>` into a `Box<[T]>`
+    ///
+    /// This conversion does not allocate on the heap and happens in place.
+    ///
+    #[unstable(feature = "box_into_boxed_slice", issue = "71582")]
+    pub fn into_boxed_slice(boxed: Box<T>) -> Box<[T]> {
+        // *mut T and *mut [T; 1] have the same size and alignment
```
Comment (shepmaster) [inline src/liballoc/boxed.rs:249]: This comment doesn't match the code. `*mut T` ≠ `*mut [T]`

Comment (cuviper) [reply src/liballoc/boxed.rs:249]: That comment is justifying the cast to an array of length 1. The slice comes after.

Comment (shepmaster) [reply src/liballoc/boxed.rs:249]: What justifies the subsequent cast?

Comment (cuviper) [reply src/liballoc/boxed.rs:249]: That's just unsizing, but we could leave that off and let the `Box<[T; 1]>` implicitly coerce unsized in the return.

**Pull request merged**